workflowcheck gets stuck forever when using templates in the workflow

## Expected Behavior

When using `workflowcheck` it should always terminate.

## Actual Behavior

When using `text/template` in the workflow code `workflowcheck` gets stuck and never terminates.

## Steps to Reproduce the Problem

See my [example repo](https://github.com/mbark/temporal-workflowcheck-error) here. I added a `Makefile` to simplify running `workflowcheck`, `make workflowcheck` will download it and run the version referenced in `tools.go` for you.

To reproduce:
  1. Run `make workflowcheck`;
  1. `workflowcheck` doesn't terminate;
  1. Comment lines 16-20 in `workflow.go`;
  2. Re-run `make workflowcheck`;
  3. It terminates successfully.

## Specifications

  - Version: `go.temporal.io/sdk/contrib/tools/workflowcheck v0.0.0-20220324161800-48b3cb957491`
  - Platform: macOS

Thanks for the reproducer! I have replicated this. Strangely I cannot replicate on Windows, only Linux. Here's what the Windows output shows:

```
c:\path\to\temporal-workflowcheck-error\main.go:20:2: templateworkflowcheck.Workflow is non-deterministic, reason: calls non-deterministic function (*text/template.Template).Execute
  (*text/template.Template).Execute is non-deterministic, reason: calls non-deterministic function (*text/template.Template).execute
    (*text/template.Template).execute is non-deterministic, reason: calls non-deterministic function (*text/template.state).walk
      (*text/template.state).walk is non-deterministic, reason: calls non-deterministic function (*text/template.state).printValue
        (*text/template.state).printValue is non-deterministic, reason: calls non-deterministic function text/template.printableValue
          text/template.printableValue is non-deterministic, reason: calls non-deterministic function reflect.PtrTo
            reflect.PtrTo is non-deterministic, reason: calls non-deterministic function (*reflect.rtype).ptrTo
              (*reflect.rtype).ptrTo is non-deterministic, reason: calls non-deterministic function (*sync.Map).LoadOrStore
                (*sync.Map).LoadOrStore is non-deterministic, reason: calls non-deterministic function (*sync.Map).dirtyLocked
                  (*sync.Map).dirtyLocked is non-deterministic, reason: iterates over map
              (*reflect.rtype).ptrTo is non-deterministic, reason: calls non-deterministic function (*sync.Map).LoadOrStore
                (*sync.Map).LoadOrStore is non-deterministic, reason: calls non-deterministic function (*sync.Map).dirtyLocked
                  (*sync.Map).dirtyLocked is non-deterministic, reason: iterates over map
          text/template.printableValue is non-deterministic, reason: calls non-deterministic function reflect.PtrTo
            reflect.PtrTo is non-deterministic, reason: calls non-deterministic function (*reflect.rtype).ptrTo
              (*reflect.rtype).ptrTo is non-deterministic, reason: calls non-deterministic function (*sync.Map).LoadOrStore
                (*sync.Map).LoadOrStore is non-deterministic, reason: calls non-deterministic function (*sync.Map).dirtyLocked
                  (*sync.Map).dirtyLocked is non-deterministic, reason: iterates over map
              (*reflect.rtype).ptrTo is non-deterministic, reason: calls non-deterministic function (*sync.Map).LoadOrStore
                (*sync.Map).LoadOrStore is non-deterministic, reason: calls non-deterministic function (*sync.Map).dirtyLocked
                  (*sync.Map).dirtyLocked is non-deterministic, reason: iterates over map
          text/template.printableValue is non-deterministic, reason: calls non-deterministic function (reflect.Value).Addr
            (reflect.Value).Addr is non-deterministic, reason: calls non-deterministic function (*reflect.rtype).ptrTo
              (*reflect.rtype).ptrTo is non-deterministic, reason: calls non-deterministic function (*sync.Map).LoadOrStore
                (*sync.Map).LoadOrStore is non-deterministic, reason: calls non-deterministic function (*sync.Map).dirtyLocked
                  (*sync.Map).dirtyLocked is non-deterministic, reason: iterates over map
              (*reflect.rtype).ptrTo is non-deterministic, reason: calls non-deterministic function (*sync.Map).LoadOrStore
                (*sync.Map).LoadOrStore is non-deterministic, reason: calls non-deterministic function (*sync.Map).dirtyLocked
                  (*sync.Map).dirtyLocked is non-deterministic, reason: iterates over map
```

I am continuing to investigate Linux. In the meantime, you can add `//workflowcheck:ignore` either above or at the end of that `.Execute` line. Also, you can upgrade your version with `go get go.temporal.io/sdk/contrib/tools/workflowcheck@latest` which does have some output improvements but it doesn't solve this issue.

From the static analyzer's POV, `Template.Execute` will be considered non-deterministic due to its internals having lots of non-deterministic code. So you're going to need to ignore `text/template` anyways with `//workflowcheck:ignore`. Technically you could make that as always deterministic via config (see README at https://github.com/temporalio/sdk-go/tree/master/contrib/tools/workflowcheck) but then if other people use the package non-deterministically you'll never know.

--- a/contrib/tools/workflowcheck/determinism/reason.go
+++ b/contrib/tools/workflowcheck/determinism/reason.go
@@ -79,28 +79,32 @@ func (n NonDeterminisms) AppendChildReasonLines(
 	includePos bool,
 	pkg *types.Package,
 	lookupCache *PackageLookupCache,
+	seenPos map[string]bool,
 ) []string {
 	for _, reason := range n {
 		reasonStr := reason.String()
-		if includePos {
-			// Relativize path if it at least starts with working dir
-			pos := reason.Pos()
-			filename := pos.Filename
-			if wd, err := os.Getwd(); err == nil && strings.HasPrefix(filename, wd) {
-				if relFilename, err := filepath.Rel(wd, filename); err == nil {
-					filename = relFilename
-				}
+		// Relativize path if it at least starts with working dir
+		pos := reason.Pos()
+		filename := pos.Filename
+		if wd, err := os.Getwd(); err == nil && strings.HasPrefix(filename, wd) {
+			if relFilename, err := filepath.Rel(wd, filename); err == nil {
+				filename = relFilename
 			}
-			reasonStr += fmt.Sprintf(" at %v:%v:%v", filename, pos.Line, pos.Column)
+		}
+		posStr := fmt.Sprintf("%v:%v:%v", filename, pos.Line, pos.Column)
+		if includePos {
+			reasonStr += " at " + posStr
 		}
 		s = append(s, fmt.Sprintf("%v is non-deterministic, reason: %v",
 			strings.Repeat(depthRepeat, depth)+subject, reasonStr))
-		// Recurse if func call
-		if funcCall, _ := reason.(*ReasonFuncCall); funcCall != nil {
+		// Recurse if func call and we haven't seen this pos str before
+		seen := seenPos[posStr]
+		seenPos[posStr] = true
+		if funcCall, _ := reason.(*ReasonFuncCall); funcCall != nil && !seen {
 			childPkg, childPkgNonDet := lookupCache.PackageNonDeterminismsFromName(pkg, funcCall.PackageName())
 			if childNonDet := childPkgNonDet[funcCall.FuncName]; len(childNonDet) > 0 {
 				s = childNonDet.AppendChildReasonLines(funcCall.FuncName, s, depth+1,
-					depthRepeat, includePos, childPkg, lookupCache)
+					depthRepeat, includePos, childPkg, lookupCache, seenPos)
 			}
 		}
 	}

--- a/contrib/tools/workflowcheck/workflow/checker.go
+++ b/contrib/tools/workflowcheck/workflow/checker.go
@@ -180,7 +180,7 @@ func (c *Checker) Run(pass *analysis.Pass) error {
 				// One report per reason
 				for _, reason := range nonDeterminisms {
 					lines := determinism.NonDeterminisms{reason}.AppendChildReasonLines(
-						fn.FullName(), nil, 0, depthRepeat, c.IncludePosOnMessage, fn.Pkg(), lookupCache)
+						fn.FullName(), nil, 0, depthRepeat, c.IncludePosOnMessage, fn.Pkg(), lookupCache, map[string]bool{})
 					pass.Report(analysis.Diagnostic{Pos: callExpr.Pos(), Message: strings.Join(lines, hierarchySeparator)})
 				}
 			}

--- a/contrib/tools/workflowcheck/workflow/testdata/src/a/workflow.go
+++ b/contrib/tools/workflowcheck/workflow/testdata/src/a/workflow.go
@@ -1,6 +1,7 @@
 package a //want package:"\\d+ non-deterministic vars/funcs"
 
 import (
+	"text/template"
 	"time"
 
 	"go.temporal.io/sdk/worker"
@@ -13,6 +14,7 @@ func PrepWorkflow() {
 	wrk.RegisterWorkflow(WorkflowCallTime)             // want "a.WorkflowCallTime is non-deterministic, reason: calls non-deterministic function time.Now"
 	wrk.RegisterWorkflow(WorkflowCallTimeTransitively) // want "a.WorkflowCallTimeTransitively is non-deterministic, reason: calls non-deterministic function a.SomeTimeCall"
 	wrk.RegisterWorkflow(WorkflowIterateMap)           // want "a.WorkflowIterateMap is non-deterministic, reason: iterates over map"
+	wrk.RegisterWorkflow(WorkflowWithTemplate)         // want "a.WorkflowWithTemplate is non-deterministic, reason: calls non-deterministic function \\(\\*text/template\\.Template\\)\\.Execute.*"
 }
 
 func WorkflowNop(ctx workflow.Context) error {
@@ -38,4 +40,8 @@ func WorkflowIterateMap(ctx workflow.Context) error { // want WorkflowIterateMap
 	for range m {
 	}
 	return nil
+}
+
+func WorkflowWithTemplate(ctx workflow.Context) error { // want WorkflowWithTemplate:"calls non-deterministic function \\(\\*text/template\\.Template\\)\\.Execute.*"
+	return template.New("mytmpl").Execute(nil, nil)
 }

